### PR TITLE
Switch wrapping to textwrap with adapter and unified prefixes

### LINKED ISSUE #80: Refactor text wrapping and prefix logic to use `textwrap` and `unicode-width` crates
## Context
The current implementation for text wrapping and prefix handling involves custom tokenization and manual logic, which increases code complexity and maintenance burden. The suggestion is to replace this with the `textwrap` and `unicode-width` crates, which are well-tested and handle Unicode, word boundaries, and hyphens more robustly.

## Proposed Improvements
1. **Replace Custom Tokenization and Wrapping:**
   - Remove the `tokenize_markdown` and `wrap_preserving_code` custom logic.
   - Use `textwrap`'s `wrap` function with appropriate options (e.g., `break_words(false)` and `UnicodeBreakWordSeparator`) to handle word boundaries and Unicode correctly.

2. **Consolidate Prefix Handling:**
   - Merge the logic of `append_wrapped_with_prefix` and `handle_prefix_line` into a single helper function (e.g., `wrap_with_prefix`).
   - Compute the available width once, wrap the text, and then prepend the prefix or indent as needed for each line.

3. **Simplify Main Loop:**
   - In the main wrapping function (e.g., `wrap_text`), call the new `wrap_with_prefix` helper instead of the old manual functions.
   - Remove all manual tokenization and buffer-flushing code.

## Benefits
- Reduces code size from ~400 lines to ~50 lines.
- Leverages well-tested libraries for Unicode and word handling.
- Maintains all existing test cases unchanged.

## Suggested Action Items
- [ ] Refactor the codebase to use `textwrap` and `unicode-width` for wrapping and prefix handling.
- [ ] Remove obsolete custom tokenization and wrapping logic.
- [ ] Ensure all existing tests pass after the refactor.
- [ ] Add/Update documentation to reflect the new implementation.

## Example Implementation
See the following code snippets for reference:

```rust
use textwrap::{wrap, Options};

fn wrap_preserving_code(text: &str, width: usize) -> Vec<String> {
    let opts = Options::new(width)
        .break_words(false)
        .word_separator(textwrap::word_separator::UnicodeBreakWordSeparator::new());
    wrap(text, &opts)
        .into_iter()
        .map(String::from)
        .collect()
}
```

```rust
use unicode_width::UnicodeWidthStr;

fn wrap_with_prefix(
    out: &mut Vec<String>,
    prefix: &str,
    text: &str,
    width: usize,
    repeat_prefix: bool,
) {
    let pw = UnicodeWidthStr::width(prefix);
    let avail = width.saturating_sub(pw).max(1);
    let lines = wrap_preserving_code(text, avail);

    for (i, line) in lines.into_iter().enumerate() {
        if i == 0 {
            out.push(format!("{prefix}{line}"));
        } else {
            let indent = if repeat_prefix {
                prefix.to_string()
            } else {
                " ".repeat(pw)
            };
            out.push(format!("{indent}{line}"));
        }
    }
}
```

---

This refactor will greatly simplify the codebase and improve maintainability. Please consider implementing these changes.

---

I created this issue for @leynos from https://github.com/leynos/mdtablefix/pull/79#discussion_r2211777533.

<details>
<summary>Tips and commands</summary>

#### Getting Help

- [Contact our support team](mailto:support@sourcery.ai) for questions or feedback.
- Visit our [documentation](https://docs.sourcery.ai) for detailed guides and information.
- Keep in touch with the Sourcery team by following us on [X/Twitter](https://x.com/SourceryAI), [LinkedIn](https://www.linkedin.com/company/sourcery-ai/) or [GitHub](https://github.com/sourcery-ai).

</details>

### PR DESCRIPTION
**Pull request by leynos**
## Summary
Switch the wrapping engine from a bespoke LineBuffer-driven path to a textwrap-based implementation, with a small adapter to preserve existing public entry points. This delivery consolidates display-width-aware prefix handling and introduces Markdown-aware fragment modeling that delegates line fitting to textwrap while preserving code spans, links, and trailing spaces.

## Changes
- Dependency updates
  - Add `textwrap` to Cargo.toml (Rust 1.89+ compatibility / 2024 edition).
  - Ensure `unicode-width` remains present and used for display-width calculations.
- Core refactor
  - Replace the bespoke LineBuffer-driven wrapping loop in the active wrap path with a textwrap-based adapter.
  - Centralize display-width-aware prefix handling into a single helper for prefixes and continuation lines.
  - Introduce a minimal adapter so existing wrap entry points (`mdtablefix::wrap::wrap_text`, `Token`, `tokenize_markdown`) continue to function for this first delivery.
- Modules touched
  - `src/wrap.rs`, `src/wrap/paragraph.rs`, and `src/wrap/inline.rs` as primary touchpoints.
  - Add/adjust an adapter layer to route line breaking through `textwrap` while preserving inline fragment semantics.
  - Introduce `src/wrap/inline/postprocess.rs` and `src/wrap/inline/test_support.rs` for post-processing and test helpers.
- Tests
  - Promote and align active wrap tests to ensure parity during the transition (updated in `src/wrap/tests.rs`, `tests/wrap_unit.rs`, `tests/wrap_cli.rs`, `tests/wrap_renumber.rs`, and related CLI/markdown tests).
  - Add regression coverage for critical wrap scenarios (code spans, links, prefixes, bullets, blockquotes, code blocks).
- Documentation
  - Update user-facing docs to describe the new wrapping approach and clarify that `textwrap` + `unicode-width` handle width-sensitive breaking.
  - Revise README.md, `docs/architecture.md`, and `docs/trailing-spaces.md` to reflect architecture changes.
  - Add new execution plan docs describing migration and constraints (e.g., `docs/execplans/replace-bespoke-wrapping-with-textwrap-and-unicode-width.md`).
- API safety
  - Public API (e.g., `Token`, `tokenize_markdown`, `wrap_text`) preserved for this first pass.
- Artifacts touched/added
  - `src/wrap.rs`, `src/wrap/paragraph.rs`, `src/wrap/inline.rs`, `src/wrap/inline/postprocess.rs`, `src/wrap/inline/test_support.rs`, `src/wrap/inline/tests.rs`, `src/wrap/tests.rs` (updated), removal of dead `src/wrap/line_buffer.rs`.
  - Documentation updates and new execplan doc file: `docs/execplans/replace-bespoke-wrapping-with-textwrap-and-unicode-width.md`.

## Rationale
- Move to a battle-tested, width-aware wrapping engine while maintaining public API stability.
- Ensure Unicode display width correctness via `unicode-width`.
- Provide a safe, incremental transition by introducing a small adapter and preserving existing entry points.

## Testing plan
- Focused test targets and general test suites:
  - cargo test -- --list | rg wrap
  - cargo test wrap_text_preserves_code_spans --lib
  - cargo test wrap_text_multiple_code_spans --lib
  - cargo test --test wrap_unit
  - cargo test --test wrap_cli
  - cargo test --test wrap_renumber
  - cargo test --test cli
  - cargo test --test cli_frontmatter
  - cargo test --test code_emphasis
  - cargo test --test markdownlint
- End-to-end validation: verify nested blockquotes, inline code, and links wrap correctly while preserving spans.

## Acceptance criteria
1. wrap_text preserves behavior covered by the active wrap tests (code spans, links, prefixes, fences, hard breaks, trailing spaces).
2. `src/wrap/paragraph.rs` uses a single, display-width-aware helper for prefix wrapping instead of a two-step approach.
3. The bespoke LineBuffer-driven wrapping loop is removed from the active wrap path; line breaking is delegated to textwrap.
4. `tokenize_markdown` remains intact and documented as out of scope for this delivery (or its broader removal is clearly approved with updated tests/docs).
5. Documentation (`README.md`, `docs/architecture.md`, and `docs/trailing-spaces.md`) accurately describe the new implementation.
6. All quality gates (fmt, lint, tests, markdownlint, nixie) pass.

## Rollback plan
If acceptance criteria are not met or regressions appear, revert the changes to the prior wrapping path and re-evaluate in a narrower scope, then reintroduce `textwrap` support in a controlled, incremental manner.

## Artifacts and notes
- Public token API and entry points remain stable for this first delivery.
- Documentation updates reflect the new wrapping approach and the minimal adapter strategy.
- Task: https://www.devboxer.com/task/204659f7-601a-4837-aa65-dbf3907e688b
- 📝 Closes #80

## Summary by Sourcery

Refactor the Markdown wrapping engine to delegate inline line-breaking to the textwrap crate with unicode-width–based fragment widths while unifying display-width-aware prefix handling, and update documentation and tests to reflect and validate the new behavior.

Enhancements:
- Replace the bespoke LineBuffer-based inline wrapping loop with a fragment adapter over textwrap::wrap_algorithms::wrap_first_fit that preserves Markdown inline code, links, whitespace, and trailing spaces.
- Centralize display-width-aware prefix and continuation handling in ParagraphWriter, ensuring consistent indentation for blockquotes, bullets, footnotes, and checkboxes.
- Clarify architecture, trailing-space semantics, and wrapping flow in documentation, including new execution plan guidance for the textwrap-based wrapping engine and minor doc style/table formatting cleanups.

Documentation:
- Document the new textwrap + unicode-width driven wrapping flow, including updated diagrams and trailing-space behavior, and add an execution plan for the wrapping refactor.

Tests:
- Expand active wrap tests to cover nested blockquotes, footnote URL continuations, and checkbox indentation while removing obsolete LineBuffer-specific tests.

Chores:
- Remove the now-unused LineBuffer module from the wrapping implementation and add the textwrap dependency to Cargo.toml.

## Task/Notes
- Task: https://www.devboxer.com/task/204659f7-601a-4837-aa65-dbf3907e688b
- 📝 Closes #80

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

### Assessment against linked issues

| Issue | Objective | Addressed | Explanation |
| --- | --- | --- | --- |
| [#80](https://github.com/leynos/mdtablefix/issues/80) | Consolidate and simplify prefix handling by introducing a single helper (e.g., `wrap_with_prefix`) that computes available width once and applies prefixes/indents for each wrapped line, and simplify the main wrapping loop to use this helper. | ❌ | No changes are made to `src/wrap/paragraph.rs`, `src/wrap.rs`, or related prefix-handling code in the PR. The added document describes an intention to consolidate prefix handling, but the implementation has not been altered accordingly. |
| [#80](https://github.com/leynos/mdtablefix/issues/80) | Update tests and documentation so that existing tests pass unchanged under the new implementation, and docs describe the new `textwrap` + `unicode-width` based wrapping engine. | ❌ | The PR does not modify any test files or existing documentation such as `README.md`, `docs/architecture.md`, or `docs/trailing-spaces.md`. It only introduces a new planning document; there is no evidence that tests were adapted or that existing docs were updated to reflect a new implementation. |

File: src/wrap/inline.rs [hunk 1]
```diff
@@ -169,99 +168,226 @@ pub(super) fn attach_punctuation_to_previous_line(
...
+                !fragment.text.chars().all(char::is_whitespace)
+                    && !is_inline_code_token(fragment.text.as_str())
+                    && !looks_like_link(fragment.text.as_str())
+            });
+
+        if should_move_atomic_tail || should_move_plain_tail {
+            let trailing_fragment = lines[index]
+                .pop()
+                .expect("line selected for tail rebalancing contains a trailing fragment");
+            lines[index + 1].insert(0, trailing_fragment);
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/wrap/inline.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

~~~
Please address the comments from this code review:

## Overall Comments
- The new `wrap_preserving_code` path repeatedly clones `buffer` into `candidate` and re-runs `wrap_first_fit` on the whole fragment list for each incoming fragment; consider restructuring to track the current line width or reusing previous wrap state so you don’t pay an O(n²) cost on long lines.
- The whitespace/atomic-span handling in `merge_whitespace_only_lines` and `rebalance_atomic_tails` is quite dense with nested conditionals and implicit invariants; factoring some of these checks into smaller, well-named helpers would make the wrapping heuristics easier to follow and maintain.

## Individual Comments

### Comment 1
<location path="src/wrap/inline.rs" line_range="365-368" />
<code_context>
+    let mut lines = Vec::new();
+    let mut buffer: Vec<InlineFragment> = Vec::new();
+
+    for fragment in fragments {
+        let mut candidate = buffer.clone();
+        candidate.push(fragment);
+        let wrapped = wrap_first_fit(&candidate, &[width_as_f64(width)]);
+        let raw_lines = wrapped.iter().map(|line| line.to_vec()).collect::<Vec<_>>();
+        let mut grouped_lines = merge_whitespace_only_lines(&raw_lines);
</code_context>
<issue_to_address>
**suggestion (performance):** Re-wrapping the entire prefix for each fragment with cloning leads to quadratic behavior and unnecessary allocations.

`wrap_preserving_code` clones `buffer` into `candidate` on every iteration and re-runs `wrap_first_fit` over the full prefix each time, making long paragraphs O(n²) in time and allocations. To improve this, consider an incremental strategy (e.g., only re-wrapping the last line or using a streaming/incremental wrapping API) or reuse a single `candidate` buffer and mutate it instead of cloning each time.

Suggested implementation:

```rust
    let fragments = build_fragments(&tokens);
    let mut lines = Vec::new();
    let mut buffer: Vec<InlineFragment> = Vec::new();

    for fragment in fragments {
        // Incrementally extend the current buffer instead of cloning it on every iteration.
        buffer.push(fragment);
        let wrapped = wrap_first_fit(&buffer, &[width_as_f64(width)]);

```

```rust
        let raw_lines = wrapped.iter().map(|line| line.to_vec()).collect::<Vec<_>>();
        let mut grouped_lines = merge_whitespace_only_lines(&raw_lines);
        rebalance_atomic_tails(&mut grouped_lines);

        // If everything still fits on a single line, keep the extended buffer as-is.
        if grouped_lines.len() == 1 {
            continue;
        }

```

This edit removes the per-iteration clone and allocation while preserving the existing wrapping behavior. It assumes the remainder of `wrap_preserving_code` already updates `buffer` to contain the fragments of the last (incomplete) line when multiple lines are produced. If that logic does not exist yet, you should:
1. After the `for line in &grouped_lines[..grouped_lines.len() - 1] { ... }` loop, set `buffer = grouped_lines.last().cloned().unwrap_or_default();` so that `buffer` only holds the fragments of the trailing line that may still accept more fragments.
</issue_to_address>

### Comment 2
<location path="src/wrap/inline.rs" line_range="171" />
<code_context>
-    if span.start >= span.end {
-        return;
-    }
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct InlineFragment {
+    text: String,
</code_context>
<issue_to_address>
**issue (complexity):** Consider adding a `FragmentKind` enum and helper methods on `InlineFragment` so fragment classification is centralized and branching in the line post‑processing functions is greatly simplified.

You can reduce a lot of the branching and duplicated logic by making `InlineFragment` carry explicit “kind” metadata and centralizing the predicates. That keeps the current behavior but makes `merge_whitespace_only_lines` and `rebalance_atomic_tails` much easier to read/maintain.

### 1. Add `FragmentKind` and helper methods

```rust
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum FragmentKind {
    Whitespace,
    InlineCode,
    Link,
    Plain,
}

#[derive(Debug, Clone, PartialEq, Eq)]
struct InlineFragment {
    text: String,
    width: usize,
    kind: FragmentKind,
}

impl InlineFragment {
    fn new(text: String) -> Self {
        let width = UnicodeWidthStr::width(text.as_str());
        let kind = if text.chars().all(char::is_whitespace) {
            FragmentKind::Whitespace
        } else if is_inline_code_token(&text) {
            FragmentKind::InlineCode
        } else if looks_like_link(&text) {
            FragmentKind::Link
        } else {
            FragmentKind::Plain
        };
        Self { text, width, kind }
    }

    fn is_whitespace(&self) -> bool {
        self.kind == FragmentKind::Whitespace
    }

    fn is_atomic(&self) -> bool {
        matches!(self.kind, FragmentKind::InlineCode | FragmentKind::Link)
    }

    fn is_plain(&self) -> bool {
        self.kind == FragmentKind::Plain
    }
}
```

You can then simplify the construction in `build_fragments` (no behavior change):

```rust
fn build_fragments(tokens: &[String]) -> Vec<InlineFragment> {
    let mut fragments = Vec::new();
    let mut i = 0;

    while i < tokens.len() {
        let (group_end, _) = determine_token_span(tokens, i);
        let span = i..group_end;
        let mut text = String::new();
        push_span_text(&mut text, tokens, span);
        fragments.push(InlineFragment::new(text));
        i = group_end;
    }

    fragments
}
```

### 2. Use `FragmentKind` in the line post‑processing

With the kind metadata, `merge_whitespace_only_lines` and `rebalance_atomic_tails` lose most of their `chars().all(char::is_whitespace)`, `is_inline_code_token`, and `looks_like_link` calls:

```rust
fn merge_whitespace_only_lines(lines: &[Vec<InlineFragment>]) -> Vec<Vec<InlineFragment>> {
    let mut merged = Vec::with_capacity(lines.len());
    let mut pending_whitespace = Vec::new();

    for (index, mut line) in lines.iter().cloned().enumerate() {
        let is_whitespace_only = line.iter().all(|f| f.is_whitespace());

        if is_whitespace_only {
            let next_starts_atomic = lines
                .get(index + 1)
                .and_then(|next| next.first())
                .is_some_and(|f| f.is_atomic());

            let line_is_single_space = line.iter().map(|f| f.text.as_str()).collect::<String>() == " ";
            let previous_line_has_single_fragment =
                merged.last().is_some_and(|prev| prev.len() == 1);

            let mut should_carry_whitespace = !line_is_single_space;

            if line_is_single_space && !next_starts_atomic {
                if let Some(previous_line) = merged.last_mut() {
                    let should_move_previous_atomic =
                        previous_line.last().is_some_and(|f| f.is_atomic());
                    if should_move_previous_atomic {
                        let previous_atomic = previous_line.pop().unwrap();
                        pending_whitespace.push(previous_atomic);
                        if previous_line.is_empty() {
                            merged.pop();
                        }
                        should_carry_whitespace = true;
                    }
                }
            }

            if line_is_single_space && previous_line_has_single_fragment {
                should_carry_whitespace = true;
            }

            if should_carry_whitespace {
                pending_whitespace.extend(line);
            }
            continue;
        }

        if pending_whitespace.is_empty() {
            merged.push(line);
        } else {
            pending_whitespace.append(&mut line);
            merged.push(std::mem::take(&mut pending_whitespace));
        }
    }

    if !pending_whitespace.is_empty() {
        if let Some(last_line) = merged.last_mut() {
            last_line.append(&mut pending_whitespace);
        } else {
            merged.push(pending_whitespace);
        }
    }

    merged
}
```

And similarly in `rebalance_atomic_tails`:

```rust
fn rebalance_atomic_tails(lines: &mut [Vec<InlineFragment>]) {
    for index in 0..lines.len().saturating_sub(1) {
        let next = &mut lines[index + 1];
        let next_starts_with_single_space =
            next.first().is_some_and(|f| f.text == " " && f.is_whitespace());
        let next_continues_with_plain_text =
            next.get(1).is_some_and(|f| f.is_plain());

        if !next_starts_with_single_space || !next_continues_with_plain_text {
            continue;
        }

        let current = &mut lines[index];
        let should_move_atomic_tail = current.last().is_some_and(|f| f.is_atomic());
        let should_move_plain_tail =
            current.len() > 1 && current.last().is_some_and(|f| f.is_plain());

        if should_move_atomic_tail || should_move_plain_tail {
            let trailing_fragment = current.pop().unwrap();
            next.insert(0, trailing_fragment);
        }
    }
}
```

This keeps your current multi‑pass behavior and all layout rules, but:

- Removes repeated classification logic from every call site.
- Reduces boolean noise by exposing higher‑level predicates (`is_whitespace`, `is_atomic`, `is_plain`) that describe intent.
- Makes it much easier to later collapse `merge_whitespace_only_lines` and `rebalance_atomic_tails` into a single “line normalizer” if you decide to, without touching the classification logic again.
</issue_to_address>
~~~

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

🚥 Pre-merge checks | ✅ 5 | ❌ 4

### ❌ Failed checks (4 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Docstring Coverage | ⚠️ Warning | Docstring coverage is 31.58% which is insufficient. The required threshold is 80.00%. | Write docstrings for the functions missing them to satisfy the coverage threshold. |
| Testing | ⚠️ Warning | New postprocess.rs module functions and InlineFragment struct lack direct unit test coverage despite implementing critical whitespace merging and classification logic. | Add dedicated unit tests for postprocess module functions, InlineFragment construction, and wrap\_with\_prefix to validate whitespace merging, tail rebalancing, and prefix handling logic. |
| User-Facing Documentation | ⚠️ Warning | PR modifies --wrap functionality but docs/users-guide.md lacks paragraph wrapping documentation whilst README.md describes wrap behaviour. | Add 'Paragraph wrapping' section to docs/users-guide.md documenting --wrap flag, hyphenation, inline code preservation, task list indentation, fenced code blocks, and Unicode display-width awareness. |
| Developer Documentation | ⚠️ Warning | Pull request introduces substantial internal API changes and architectural refactoring without adequate documentation in developers-guide.md as required by custom check. | Create ADR documenting textwrap refactoring decision rationale; expand developers-guide.md with dedicated wrap module architecture section detailing new structures, dependencies, and design constraints. |

**Comment by leynos** [comment]
@coderabbitai pause

File: docs/developers-guide.md [hunk 2]
```diff
@@ -88,3 +88,50 @@ The rationale for the staged table reflow pipeline is recorded in
...
+   width (measured with `unicode-width`) and a `FragmentKind` tag
+   (`Whitespace`, `InlineCode`, `Link`, or `Plain`) computed once at
+   construction time by `classify_fragment`.
+
+3. **Line fitting and post-processing.** `wrap_preserving_code` calls
+   `textwrap::wrap_algorithms::wrap_first_fit` over the accumulated fragment
+   buffer, then applies `merge_whitespace_only_lines` and
+   `rebalance_atomic_tails` from `src/wrap/inline/postprocess.rs` to normalise
+   whitespace-only wrap lines and rebalance atomic tails that would otherwise
+   appear isolated at the start of continuation lines.
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/developers-guide.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

### ❌ Failed checks (2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing | ⚠️ Warning | The `wrap_preserving_code` function lacks direct unit tests, relying only on indirect testing through integration tests and helper function tests. | Add direct unit tests for `wrap_preserving_code` covering basic fitting, multiline wrapping, empty input, code span preservation, Markdown links, hard breaks, and width compliance assertions. |
| Module-Level Documentation | ⚠️ Warning | pub fn tokenize\_markdown() lacks any documentation, and pub use fence::... re-export on src/wrap.rs line 20 lacks a docstring. | Add comprehensive doc comments before tokenize\_markdown() describing its purpose, parameters, and return value, and before the pub use statement explaining FenceTracker and is\_fence roles. |
